code/go-utils/sonarlog/ustr.go

```diff
@@ -140,3 +202,127 @@ func NewUstrFacade() *UstrFacade {
 func (uf *UstrFacade) Alloc(s string) Ustr {
 	return StringToUstr(s)
 }
+
+func (uf *UstrFacade) AllocBytes(bs []byte) Ustr {
+	return BytesToUstr(bs)
+}
+
+// hashtable maps a string or []byte to a hashnode, treating the two key types as equivalent.  The
+// node carries the name and the Ustr value of that name.
+
+const (
+	// These matter.  L=10 is pretty aggressive.
+	inverseLoad     uint32 = 10
+	initialCapacity uint32 = 93
+)
+
+type hashtable struct {
+	table     []*hashnode
+	size      uint32
+	divisor   uint32
+	remaining uint32
+}
+
+type hashnode struct {
+	hash hashcode
+	name string
+	ustr Ustr
+	next *hashnode
+}
+
+func newHashtable() hashtable {
+	size := inverseLoad * initialCapacity
+	return hashtable{
+		table:     make([]*hashnode, size),
+		size:      0,
+		divisor:   size,
+		remaining: initialCapacity,
+	}
+}
+
+func (ht *hashtable) getString(h hashcode, s string) *hashnode {
+	slot := uint32(h) % ht.divisor
+	probe := ht.table[slot]
+	for probe != nil && s != probe.name {
+		probe = probe.next
+	}
+	return probe
+}
+
+func (ht *hashtable) getBytes(h hashcode, bs []byte) *hashnode {
+	slot := uint32(h) % ht.divisor
+	probe := ht.table[slot]
+	for probe != nil {
+		name := probe.name
+		found := true
+		if len(name) != len(bs) {
+			found = false
+		} else {
+			for i := range name {
+				if name[i] != bs[i] {
+					found = false
+					break
+				}
+			}
+		}
+		if found {
+			break
+		}
+		probe = probe.next
+	}
+	return probe
+}
+
+func (ht *hashtable) insert(h hashcode, name string, ustr Ustr) {
+	ht.maybeRehash()
+	ht.remaining--
+	ht.size++
+	slot := uint32(h) % ht.divisor
+	node := &hashnode{
+		hash: h,
+		ustr: ustr,
+		name: name,
+		next: ht.table[slot],
+	}
+	ht.table[slot] = node
+}
+
+func (ht *hashtable) maybeRehash() {
+	if ht.remaining == 0 {
+		newSize := 2 * uint32(len(ht.table))
+		newRemaining := uint32(len(ht.table)) / inverseLoad
+		newTable := make([]*hashnode, newSize)
+		for _, l := range ht.table {
+			for l != nil {
+				p := l
+				l = l.next
+				slot := uint32(p.hash) % newSize
+				p.next = newTable[slot]
+				newTable[slot] = p
+			}
+		}
+		ht.table = newTable
+		ht.divisor = newSize
+		ht.remaining = newRemaining
+	}
+}
+
+// hashString and hashBytes must return the same values for the same bytes in the same order.
+
+type hashcode uint32
+
+func hashString(s string) hashcode {
+	h := uint32(0)
+	for i := range s {
+		h = (h << 4) + uint32(s[i]) + (h >> 28)
```

Comment:
Don't know where this algorithm comes from.  It would be useful to document / do a little research.

Reply:
Indeed, hashString is broken, as it iterates on runes while hashBytes iterates on bytes - they will diverge for non-ascii content.

Reply:
Fixed.

Reply:
I will file a followup bug to analyze performance and load of the hash table.